run_tests.py

```diff
@@ -80,39 +88,146 @@ def decompile_and_compare(
+def decompile_and_capture_output(options: Options, brief_crashes: bool) -> str:
     out_string = io.StringIO()
     with contextlib.redirect_stdout(out_string):
         returncode = decompile(options)
+    out_text = out_string.getvalue()
+    # Rewrite paths in the output to be relative (e.g. in tracebacks)
+    out_text = out_text.replace(str(Path(__file__).parent), ".")
     if returncode == 0:
-        return out_string.getvalue()
+        return out_text
     else:
-        return CRASH_STRING
+        if brief_crashes:
+            return CRASH_STRING
+        else:
+            return f"{CRASH_STRING}\n{out_text}"
 
 
 def run_e2e_test(
-    e2e_top_dir: Path, e2e_test_path: Path, should_overwrite: bool, coverage: Any
+    e2e_top_dir: Path,
+    e2e_test_path: Path,
+    should_overwrite: bool,
+    filter_regex: Optional[str],
+    coverage: Any,
 ) -> bool:
-    logging.info(f"Running test: {e2e_test_path.name}")
 
     ret = True
     for asm_file_path in e2e_test_path.glob("*.s"):
         old_output_path = asm_file_path.parent.joinpath(asm_file_path.stem + "-out.c")
         flags_path = asm_file_path.parent.joinpath(asm_file_path.stem + "-flags.txt")
+
+        name = f"e2e:{asm_file_path.relative_to(e2e_top_dir)}"
+        if filter_regex is not None and not re.search(filter_regex, name):
+            continue
         if coverage:
-            coverage.switch_context(str(asm_file_path.relative_to(e2e_top_dir)))
+            coverage.switch_context(name)
+        logging.info(f"Running test: {name}")
+
         if not decompile_and_compare(
-            asm_file_path, old_output_path, flags_path, should_overwrite
+            asm_file_path,
+            old_output_path,
+            brief_crashes=True,
+            should_overwrite=should_overwrite,
+            flags_path=flags_path,
+            flags=["test"],
         ):
             ret = False
     return ret
 
 
-def main(should_overwrite: bool, coverage: Any) -> int:
+def run_project_tests(
+    base_dir: Path,
+    output_dir: Path,
+    should_overwrite: bool,
+    filter_regex: Optional[str],
+    name_prefix: str,
+    coverage: Any,
+) -> bool:
+    ret = True
+    asm_dir = base_dir / "asm"
+    context_file = base_dir / "ctx.c"
+    for asm_file in asm_dir.rglob("*"):
+        if asm_file.suffix not in (".asm", ".s"):
+            continue
+        if "non_matching" in str(asm_file):
+            continue
+
+        asm_name = asm_file.name
+        if (
+            asm_name.startswith("code_data")
+            or asm_name.startswith("code_rodata")
+            or asm_name.startswith("boot_data")
+            or asm_name.startswith("boot_rodata")
+            or asm_name.endswith("_data.asm")
+            or asm_name.endswith("_rodata.asm")
+        ):
+            continue
+
+        has_context = context_file.exists()
+        flags = []
+        if has_context:
+            flags.extend(["--context", str(context_file)])
```

Comment:
yeah this repeated parsing seems disastrous for performance. possible suggestions:
- a `--cache-context-for-tests` flag with `help=argparse.SUPPRESS` that globally caches the typemap if passed (with the context filename as key, or clearing the cache from here between tests; it's okay if it's ugly)
- passing an optional `typemap` through to main.run

Reply:
I had another idea recently - put the caching in `build_typemap`: if you call `build_typemap` twice in a row with the same string, it could return the cached `TypeMap`?

Reply:
I think it's probably worthwhile to avoid repeated file reads -- context files can be pretty big

(also I'm slightly unhappy about making this non-threadsafe outside of the "anything goes" invoked-from-test case, not that it would actually come up in practice)

Reply:
I was thinking of just using [`functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache), which is threadsafe. 

The context files *are* pretty big (a few MB), I think in practice it's not that bad.

Reply:
can we lru_cache a function that takes in a filename and returns a typemap?

Reply:
Yeah - we could do that. I was a bit afraid of having the contents of the file change though. (ex: I'm not sure how the website is implemented)

I tried `lru_cache(maxsize=4)` with a ~2MB context and it only takes ~3ms to read the file & load it from the cache. That seems worth it to me to avoid some potential cache invalidation issues?
```python
In [8]: tm = src.c_types.build_typemap(open("../mm/ctx.c").read())                                                             

In [9]: %timeit src.c_types.build_typemap(open("../mm/ctx.c").read())                                                          
2.73 ms ± 41.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
Huh, that's surprisingly fast. I guess it would be fine then (if still maybe not personal preference). The website runs the script repeatedly through subprocess fwiw, without any persistence -- one reason being to deal gracefully with infinite loops.